sno/sno_repo.py

```diff
@@ -0,0 +1,146 @@
+            HEAD
+            ...      # objects, refs, etc
+    Converts it to a sno repo with the following structure:
+    repo_root_path/  # also worktree path, as before
+        .git         # contains "gitdir: .sno"
+        .sno/
+            HEAD
+            index    # Locked using .sno extension
+            ...      # etc
+    """
+
+    # Note: pygit2 can read the .sno directory okay, but it won't create it, so we do it ourselves using Path.rename()
+    # Note: We could also call `git init --separate-git-dir=.sno` to crete it but it doesn't work very well - it
+    # converts all paths to be absolute, which would mean sno repos wouldn't be renameable.
+
+    # Move .git to .sno but add a reference at .git so git can find it.
+    dot_git_path = repo_root_path / ".git"
+    assert dot_git_path.is_dir()
+    dot_sno_path = repo_root_path / ".sno"
+    assert not dot_sno_path.exists()
+
+    dot_git_path.rename(dot_sno_path)
+    dot_git_path.write_text("gitdir: .sno\n", encoding="utf-8")
+
+    repo = pygit2.Repository(str(repo_root_path.resolve()))
+    assert is_sno_repo(repo)
+
+    # Disable git commands that would mess things up:
+    repo_files.write_repo_file(repo, repo_files.INDEX, LOCKED_EMPTY_GIT_INDEX)
+
+    # Force writing to reflogs.
+    repo.config["core.logAllRefUpdates"] = "always"
+    return repo
+
+
+def is_sno_repo(pygit_repo):
+    if pygit_repo is not None:
+        # Older sno repos were just bare repos, newer sno repos are non-bare git repos that use ".sno" instead of ".git"
+        return pygit_repo.is_bare or Path(pygit_repo.path).stem == ".sno"
+    return False
+
+
+def get_sno_workdir(pygit_repo):
+    # Older sno repos are bare and use repo.path as the "workdir" - that is, where working copies are written -
+    # Newer ones have an actual workdir.
+    return Path(pygit_repo.path) if pygit_repo.is_bare else Path(pygit_repo.workdir)
+
+
+@contextlib.contextmanager
+def unlocked_git_index(pygit_repo):
+    """
+    A sno repo has an index file containing a required ".sno" extension that git doesn't support, which stops git from
+    performing various operations that we want to be only performed by sno - eg `sno checkout` does what the user wants,
+    `git checkout` would not do what the user wants, so we disable it in case the user types it out of habit.
+
+    To temporarily re-enable some of these operations from within sno, we can temporarily unlock the index by removing
+    the required ".sno" extension.
+    """
+    index_path = repo_files.repo_file_path(pygit_repo, repo_files.INDEX)
+    if not index_path.exists():
+        yield
+        return
+
+    _remove_locked_sno_extension(index_path)
+    try:
+        yield
+    finally:
+        _add_locked_sno_extension(index_path)
+
+
+def _remove_locked_sno_extension(index_path):
+    index_contents = index_path.read_bytes()
+    index_contents = index_contents[:-20]  # Trim off checksum
+    if index_contents[-8:] == LOCKED_SNO_EXTENSION:
```

Comment:
Should we be logging a warning if it's not how we expect?

Reply:
Done